### PR TITLE
Std search module snippets

### PR DESCRIPTION
**Pull request by joshuak94**

### CODE REVIEW

File: test/snippet/search/bi_fm_index_cursor_cycle.cpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+    seqan3::debug_stream << "Example cycle_back() and cycle_front()\n";
+
+    std::vector<seqan3::dna4> genome{"GAATTAATGAAC"_dna4};
+    seqan3::bi_fm_index index{genome};                              // build the bidirectional index
+
+    auto cur = index.begin();                               // create a cursor
+    // cur.cycle_back();                                    // cycle_back / cycle_front on begin() is undefined behaviour!
+    cur.extend_right("AAC"_dna4);                           // search the sequence "AAC"
+    seqan3::debug_stream << cur.path_label(genome) << '\n';         // outputs "AAC"
```
Comment (eseiler) [inline test/snippet/search/bi_fm_index_cursor_cycle.cpp]: indentation (whole file)

File: test/snippet/search/bi_fm_index_cursor_cycle.cpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
+#include <vector>
```
Comment (eseiler) [inline test/snippet/search/bi_fm_index_cursor_cycle.cpp]: should be redundant?
otherwise add newline
(same for other files)

File: test/snippet/search/bi_fm_index_cursor_extend_left_seq.cpp [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#include <vector>
+#include <seqan3/core/debug_stream.hpp>
+#include <seqan3/search/fm_index/bi_fm_index.hpp>
+#include <seqan3/alphabet/nucleotide/dna4.hpp>
```
Comment (eseiler) [inline test/snippet/search/bi_fm_index_cursor_extend_left_seq.cpp]: lexicographic order
(also check other files)

File: test/snippet/search/fm_index_cursor.cpp [hunk 4]
```diff
@@ -2,31 +2,26 @@
...
-debug_stream << cur.path_label(genome) << '\n';         // outputs "AAT"
-debug_stream << cur.last_rank() << '\n';                // outputs 3
-
-cur.cycle_back();                                       // "cur" doesn't change because the rightmost char is already the largest dna4 char.
-debug_stream << cur.path_label(genome) << '\n';         // outputs "AAT"
-debug_stream << cur.last_rank() << '\n';                // outputs 3
-//! [cycle]
-
-return 0;
+    // TODO(comments): outputs [A, A, G] instead of AAG
```
Comment (eseiler) [inline test/snippet/search/fm_index_cursor.cpp]: remove todo

File: test/snippet/std/view/subrange.cpp [hunk 5]
```diff
@@ -3,15 +3,13 @@
...
 {
-//! [example]
-dna4_vector s{"ACTTTGATAA"_dna4};
-using iterator = dna4_vector::iterator;
-auto v1 = std::ranges::subrange<iterator, iterator>{begin(s) + 2, end(s)} | view::to_char; // == "TTTGATAA"
-//! [example]
+    using seqan3::operator""_dna4;
+    seqan3::dna4_vector s{"ACTTTGATAA"_dna4};
+    using iterator = seqan3::dna4_vector::iterator;
+    auto v1 = std::ranges::subrange<iterator, iterator>{seqan3::begin(s) + 2, seqan3::end(s)} | seqan3::view::to_char;
```
Comment (eseiler) [inline test/snippet/std/view/subrange.cpp]:
```suggestion
    auto v1 = std::ranges::subrange<iterator, iterator>{std::ranges::begin(s) + 2, std::ranges::end(s)} | seqan3::view::to_char;
```

File: test/snippet/std/view/subrange.cpp [hunk 6]
```diff
@@ -3,15 +3,14 @@
...
 {
-//! [example]
-dna4_vector s{"ACTTTGATAA"_dna4};
-using iterator = dna4_vector::iterator;
-auto v1 = std::ranges::subrange<iterator, iterator>{begin(s) + 2, end(s)} | view::to_char; // == "TTTGATAA"
-//! [example]
+    using seqan3::operator""_dna4;
+    seqan3::dna4_vector s{"ACTTTGATAA"_dna4};
+    using iterator = seqan3::dna4_vector::iterator;
+    auto v1 = std::ranges::subrange<iterator, iterator>{std::ranges::begin(s) + 2, std::ranges::end(s)} |
```
Comment (eseiler) [inline test/snippet/std/view/subrange.cpp]: the `|` should be aligned to the `=`

File: test/snippet/search/bi_fm_index.cpp [hunk 7]
```diff
@@ -1,21 +1,19 @@
-#include <vector>
```
Comment (h-2) [inline test/snippet/search/bi_fm_index.cpp]: why was this removed? What about dna4 ?

File: test/snippet/search/configuration_default.cpp [hunk 8]
```diff
@@ -2,15 +2,12 @@
...
-                                  | search_cfg::output{search_cfg::text_position}
-                                  | search_cfg::mode{search_cfg::all};
-//![default]
+    // No errors, all hits as text position
+    seqan3::configuration const default_cfg = seqan3::search_cfg::max_error{seqan3::search_cfg::total{0},
+                                                                            seqan3::search_cfg::substitution{0},
+                                                                            seqan3::search_cfg::insertion{0},
+                                                                            seqan3::search_cfg::deletion{0}}
+                                            | seqan3::search_cfg::output{seqan3::search_cfg::text_position}
+                                            | seqan3::search_cfg::mode{seqan3::search_cfg::all};
```
Comment (h-2) [inline test/snippet/search/configuration_default.cpp]: for everything that isn't a view, the `|` goes on the end of line I think... @smehringer do you recall the decision?

Comment (eseiler) [reply test/snippet/search/configuration_default.cpp]: https://github.com/seqan/seqan3/wiki/General-formatting says at the end if it's not a range, otherwise (context=range) align with previous `|` or `=`

File: test/snippet/search/bi_fm_index.cpp (hunk 7 above)
Comment (eseiler) [reply test/snippet/search/bi_fm_index.cpp]: `#include <seqan3/search/fm_index/all.hpp>`
includes this already. 
Even only including `fm_index.hpp` would include the concept, which includes dna4 .. and vector  is included somewhere on the way.

Comment (joshuak94) [reply test/snippet/search/bi_fm_index.cpp]: Hmmm I completely missed dna4...

I wouldn't think `#include <vector>` is necessary, since all of the nucleotide/aminoacid alphabets come with `alphabet_vector` aliases anyways, and they all already have the vector include.

File: test/snippet/search/configuration_default.cpp (hunk 8 above)
Comment (joshuak94) [reply test/snippet/search/configuration_default.cpp]: So should I keep this or change it?

File: test/snippet/search/bi_fm_index.cpp (hunk 7 above)
Comment (joshuak94) [reply test/snippet/search/bi_fm_index.cpp]: Is the verdict now to include both dna4 and vector, or is it enough to just include dna4?

Comment (h-2) [reply test/snippet/search/bi_fm_index.cpp]: Include both! I would actually like to get rid of the vector includes in the alphabets somehow...

File: test/snippet/search/configuration_default.cpp (hunk 8 above)
Comment (h-2) [reply test/snippet/search/configuration_default.cpp]: Move `|` to the end!

File: test/snippet/search/bi_fm_index.cpp (hunk 7 above)
Comment (eseiler) [reply test/snippet/search/bi_fm_index.cpp]: Actually you need neither dna nor vector since the bi_fm_index includes them anyway... that's kinda my point.
Why include redundant stuff.

Comment (joshuak94) [reply test/snippet/search/bi_fm_index.cpp]: Okay, I can include both.

Although I think it isn't very intuitive for people to include vector when they'll be using almost exclusively `seqan3::dna4_vector foo` and not `std::vector<seqan3::dna4> foo`. I would probably assume that `seqan3::dna4_vector` implies that I only need the dna4 include.

Comment (joshuak94) [reply test/snippet/search/bi_fm_index.cpp]: (Although I now see that in these snippets we are using `std::vector<seqan3::dna4>` and not `seqan3::dna4_vector`.

File: test/snippet/search/bi_fm_index_cursor_cycle.cpp [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    // cur.cycle_front();                                   // undefined behaviour! only cycle_back() is allowed after extend_right()
+    cur.cycle_back();                                       // search the sequence "AAT"
+    seqan3::debug_stream << cur.path_label(genome) << '\n'; // outputs "AAT"
+    seqan3::debug_stream << cur.last_rank() << '\n';        // outputs 3
+
+    cur.extend_left('G'_dna4);                              // search the sequence "GAAT"
+    seqan3::debug_stream << cur.path_label(genome) << '\n'; // outputs "GAAC"
+    seqan3::debug_stream << cur.last_rank() << '\n';        // outputs 2
+
+    // cur.cycle_back();                                    // undefined behaviour! only cycle_front() is allowed after extend_left()
```
Comment (h-2) [inline test/snippet/search/bi_fm_index_cursor_cycle.cpp]: Not part of this PR, but do we really want this kind of UB? Shouldn't we throw instead?

File: test/snippet/search/configuration_default.cpp [hunk 10]
```diff
@@ -2,15 +2,12 @@
...
-                                  | search_cfg::output{search_cfg::text_position}
-                                  | search_cfg::mode{search_cfg::all};
-//![default]
+    // No errors, all hits as text position
+    seqan3::configuration const default_cfg = seqan3::search_cfg::max_error{seqan3::search_cfg::total{0},
+                                                                            seqan3::search_cfg::substitution{0},
+                                                                            seqan3::search_cfg::insertion{0},
+                                                                            seqan3::search_cfg::deletion{0}} |
+                                                                            seqan3::search_cfg::output{seqan3::search_cfg::text_position} | 
+                                                                            seqan3::search_cfg::mode{seqan3::search_cfg::all};
```
Comment (h-2) [inline test/snippet/search/configuration_default.cpp]: Now the alignment of the last two rows got broken...

File: test/snippet/search/configuration_output.cpp [hunk 11]
```diff
@@ -2,19 +2,17 @@
...
-    configuration const cfg1 = search_cfg::max_error{search_cfg::total{1},
-                                                     search_cfg::substitution{0},
-                                                     search_cfg::insertion{1},
-                                                     search_cfg::deletion{1}}
-                               | search_cfg::output{search_cfg::text_position};
+    seqan3::configuration const cfg1 = seqan3::search_cfg::max_error{seqan3::search_cfg::total{1},
+                                                                     seqan3::search_cfg::substitution{0},
+                                                                     seqan3::search_cfg::insertion{1},
+                                                                     seqan3::search_cfg::deletion{1}} |
+                                                                     seqan3::search_cfg::output{seqan3::search_cfg::text_position};
```
Comment (h-2) [inline test/snippet/search/configuration_output.cpp]: here as well

File: test/snippet/search/configuration_output.cpp [hunk 12]
```diff
@@ -2,19 +2,17 @@
...
 
     // Return cursors of the index.
-    configuration const cfg2 = search_cfg::max_error{search_cfg::substitution{0},
-                                                     search_cfg::insertion{1},
-                                                     search_cfg::deletion{1}}
-                               | search_cfg::output{search_cfg::index_cursor};
+    seqan3::configuration const cfg2 = seqan3::search_cfg::max_error{seqan3::search_cfg::substitution{0},
+                                                                     seqan3::search_cfg::insertion{1},
+                                                                     seqan3::search_cfg::deletion{1}} |
+                                                                     seqan3::search_cfg::output{seqan3::search_cfg::index_cursor};
```
Comment (h-2) [inline test/snippet/search/configuration_output.cpp]: here as well (output is not argument to max_error, but to the config itself!)

File: test/snippet/search/bi_fm_index.cpp (hunk 7 above)
Comment (h-2) [reply test/snippet/search/bi_fm_index.cpp]: @eseiler Sorry, I previously didn't see your comment or I wouldn't have responded like that.
I still think its correct to fully specify the includes. Unless FM_index is *specified* as providing dna5 you cannot rely on it (changing includes would be a breaking change, see also our stability page).

BTW: why *does* it include dna5?

File: test/snippet/search/configuration_modes.cpp [hunk 13]
```diff
@@ -2,19 +2,17 @@
...
 
     // Report all hits with best + 1 error. E.g., if the best hit has 1 error, all hits with 1 and 2 error are reported.
-    configuration const cfg2 = search_cfg::max_error{search_cfg::substitution{0},
-                                                     search_cfg::insertion{1},
-                                                     search_cfg::deletion{1}}
-                               | search_cfg::mode{search_cfg::strata{1}};
+    seqan3::configuration const cfg2 = seqan3::search_cfg::max_error{seqan3::search_cfg::substitution{0},
+                                                                     seqan3::search_cfg::insertion{1},
+                                                                     seqan3::search_cfg::deletion{1}} |
+                                                                     seqan3::search_cfg::mode{seqan3::search_cfg::strata{1}};
```
Comment (h-2) [inline test/snippet/search/configuration_modes.cpp]: still not aligned correctly

File: test/snippet/search/configuration_modes.cpp [hunk 14]
```diff
@@ -2,19 +2,17 @@
...
-    configuration const cfg1 = search_cfg::max_error{search_cfg::total{1},
-                                                     search_cfg::substitution{0},
-                                                     search_cfg::insertion{1},
-                                                     search_cfg::deletion{1}}
-                               | search_cfg::mode{search_cfg::best};
+    seqan3::configuration const cfg1 = seqan3::search_cfg::max_error{seqan3::search_cfg::total{1},
+                                                                     seqan3::search_cfg::substitution{0},
+                                                                     seqan3::search_cfg::insertion{1},
+                                                                     seqan3::search_cfg::deletion{1}} |
+                                                                     seqan3::search_cfg::mode{seqan3::search_cfg::best};
```
Comment (h-2) [inline test/snippet/search/configuration_modes.cpp]: here as well

**Comment by h-2** [comment]
please remember to name your commits according to our contributor guidelines!

File: test/snippet/search/bi_fm_index.cpp (hunk 7 above)
Comment (eseiler) [reply test/snippet/search/bi_fm_index.cpp]: Ok, then lets do `vector` and `dna4` as include.

> BTW: why _does_ it include dna5?

(dna4, right?) It's needed for a concept check, this could probably be replaced by `char`, though...
